Plugins/Public/docking_modules/Utilities.cpp

```diff
@@ -0,0 +1,300 @@
+}
+
+// Get proxy base ID for specific carrier.
+uint GetProxyBaseForCarrier(uint carrierClientID)
+{
+	uint iBaseID;
+	string systemName = wstos(HkGetSystemNickByID(Players[carrierClientID].iSystemID));
+	string proxyBaseSuffix = Watcher.Cache[carrierClientID].dockingTraits.proxyBaseSuffix;
+	string proxyBase = systemName + proxyBaseSuffix;
+	pub::GetBaseID(iBaseID, proxyBase.c_str());
+
+	return iBaseID;
+}
+
+// Get proxy base ID for specific system.
+uint GetProxyBaseForSystem(uint carrierClientID, uint iSystemID)
+{
+	uint iBaseID;
+	string systemName = wstos(HkGetSystemNickByID(iSystemID));
+	string proxyBaseSuffix = Watcher.Cache[carrierClientID].dockingTraits.proxyBaseSuffix;
+	string proxyBase = systemName + proxyBaseSuffix;
+	pub::GetBaseID(iBaseID, proxyBase.c_str());
+
+	return iBaseID;
+}
+
+// Gets character name from charfile and account.
+wstring HkGetCharnameFromCharFile(string const &charFile, CAccount *acc)
+{
+	string path;
+	path.reserve(dataPath.size() + 45);
+
+	// Get account directory name.
+	_GetFLName GetFLName = (_GetFLName)((char*)hModServer + 0x66370);
+	char szDir[12];
+	GetFLName(szDir, acc->wszAccID);
+
+	path = dataPath + "\\Accts\\MultiPlayer\\" + szDir + "\\" + charFile + ".fl";
+
+	map<string, vector<string>> variables;
+	variables["name"] = vector<string>();
+	ReadFLFile(variables, path);
+
+	wstring charname;
+	if (!variables["name"].empty())
+		charname = EncodeWStringFromStringOfBytes(variables["name"][0]);
+
+	return charname;
+}
+
+// Get path to original FL account for this charname.
+string GetFLAccPath(wstring &charname)
+{
+	string path;
+	path.reserve(dataPath.size() + 45);
+
+	char datapath[MAX_PATH];
+	GetUserDataPath(datapath);
+
+	CAccount *acc = HkGetAccountByCharname(charname);
+
+	// Get account directory name.
+	_GetFLName GetFLName = (_GetFLName)((char*)hModServer + 0x66370);
+	char szDir[12];
+	GetFLName(szDir, acc->wszAccID);
+
+	// Get character file name.
+	char szFile[12];
+	GetFLName(szFile, charname.c_str());
+
+	path = dataPath + "\\Accts\\MultiPlayer\\" + szDir + "\\" + szFile + ".fl";
+	return path;
+}
+
+// Removes all specifil lines from file, adds specific lines to end of file, or replaces one line with few another if need to insert lines into specific position. Returns result of check if file exists.
+void EditFLFile(vector<string> &linesToDelete, map<string, vector<string>> &linesToReplace, vector<string> &hookExtLinesToAdd, vector<string> &hookExtLinesToDelete, string &path)
+{
+	string tempPath = path + "-temp";
+	ofstream newFile(tempPath);
+	ifstream file(path);
+
+	// Copy all lines to new file with some editions.
+	if (file.is_open())
+	{
+		string line;
+
+	Begin:
+		while (getline(file, line))
+		{
+			// Do not scan for hookext data.
+			if (boost::algorithm::starts_with(line, "[flhook]"))
+			{
+				newFile << line << endl;
+				break;
+			}
+
+			for (string &del : linesToDelete)
+				if (boost::algorithm::starts_with(line, del))
+					goto Begin;

```

Comment:
oh no, goto. this feels like a bug waiting to happen

Reply:
Best way to break out of nested loops.

Reply:
I'd rather we did it the usual way.